src/modules/musig/main_impl.h

```diff
@@ -0,0 +1,566 @@
+    unsigned char combined_ser[33];
+    size_t combined_ser_size = sizeof(combined_ser);
+    int overflow;
+    secp256k1_scalar x;
+    secp256k1_scalar y;
+    secp256k1_sha256 sha;
+    secp256k1_gej rj;
+    secp256k1_ge rp;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(secp256k1_ecmult_gen_context_is_built(&ctx->ecmult_gen_ctx));
+    ARG_CHECK(session != NULL);
+    ARG_CHECK(nonce_commitment32 != NULL);
+    ARG_CHECK(session_id32 != NULL);
+    ARG_CHECK(combined_pk != NULL);
+    ARG_CHECK(pk_hash32 != NULL);
+    ARG_CHECK(seckey != NULL);
+
+    if (msg32 != NULL) {
+        memcpy(session->msg, msg32, 32);
+        session->msg_is_set = 1;
+    } else {
+        session->msg_is_set = 0;
+    }
+    memcpy(&session->combined_pk, combined_pk, sizeof(*combined_pk));
+    memcpy(session->pk_hash, pk_hash32, 32);
+    session->nonce_is_set = 0;
+    session->has_secret_data = 1;
+    if (n_signers == 0 || my_index >= n_signers) {
+        return 0;
+    }
+    session->n_signers = n_signers;
+    session->nonce_commitments_hash_is_set = 0;
+
+    /* Compute secret key */
+    secp256k1_scalar_set_b32(&x, seckey, &overflow);
+    if (overflow) {
+        return 0;
+    }
+    secp256k1_musig_coefficient(&y, pk_hash32, my_index);
+    secp256k1_scalar_mul(&x, &x, &y);
+    secp256k1_scalar_get_b32(session->seckey, &x);
+
+    /* Compute secret nonce */
+    secp256k1_sha256_initialize(&sha);
+    secp256k1_sha256_write(&sha, seckey, 32);
+    secp256k1_sha256_write(&sha, session_id32, 32);
+    if (session->msg_is_set) {
+        secp256k1_sha256_write(&sha, msg32, 32);
+    }
+    secp256k1_ec_pubkey_serialize(ctx, combined_ser, &combined_ser_size, combined_pk, SECP256K1_EC_COMPRESSED);
+    secp256k1_sha256_write(&sha, combined_ser, combined_ser_size);
+    secp256k1_sha256_finalize(&sha, session->secnonce);
+    secp256k1_scalar_set_b32(&x, session->secnonce, &overflow);
+    if (overflow) {
+        return 0;
+    }
+
+    /* Compute public nonce and commitment */
+    secp256k1_ecmult_gen(&ctx->ecmult_gen_ctx, &rj, &x);
+    secp256k1_ge_set_gej(&rp, &rj);
+    secp256k1_pubkey_save(&session->nonce, &rp);
+
+    if (nonce_commitment32 != NULL) {
+        unsigned char commit[33];
+        size_t commit_size = sizeof(commit);
+        secp256k1_sha256_initialize(&sha);
+        secp256k1_ec_pubkey_serialize(ctx, commit, &commit_size, &session->nonce, SECP256K1_EC_COMPRESSED);
+        secp256k1_sha256_write(&sha, commit, commit_size);
+        secp256k1_sha256_finalize(&sha, nonce_commitment32);
+    }
+
+    return 1;
+}
+
+
+int secp256k1_musig_session_get_public_nonce(const secp256k1_context* ctx, secp256k1_musig_session *session, secp256k1_pubkey *nonce, secp256k1_musig_signer_data *signers, const unsigned char *const *commitments, size_t n_commitments) {
```

Comment:
Interesting that all the failure modes here come from API misuse. I wonder if it's reasonable to turn some of these `return 0`s into `ARG_CHECK`s?

My feeling is no, I prefer to use `ARG_CHECK` only for things that could be prevented with Rust's type system, but I could go either way.

Reply:
Why would it be better to have an `ARG_CHECK` exactly for that?
In general for me it's a bit difficult to imagine what exactly rusts type system can prevent, but it won't be a big problem switching from `return 0` to `ARG_CHECK` once we have the rust wrapper.

Reply:
My thinking is that users can assume `ARG_CHECK` failures never happen unless they're misusing the API, while "normal" failures are something they should check for.